src/Sulu/Bundle/CategoryBundle/Category/CategoryManager.php

```diff
@@ -235,246 +236,187 @@ public function getFieldDescriptors($locale)
+        }
+
+        return $this->getApiObject($entity, $locale);
     }
 
     /**
-     * Returns a category with a given id.
-     *
-     * @param int $id the id of the category
-     *
-     * @return CategoryEntity
+     * {@inheritdoc}
      */
-    public function findById($id)
+    public function findByIds(array $ids, $locale)
     {
-        return $this->categoryRepository->findCategoryById($id);
+        $entities = $this->categoryRepository->findCategoriesByIds($ids);
+
+        return $this->getApiObjects($entities, $locale);
     }
 
     /**
-     * Returns a category with a given key.
-     *
-     * @param string $key the key of the category
-     *
-     * @return CategoryEntity
+     * {@inheritdoc}
      */
-    public function findByKey($key)
+    public function findBetween($fromIds, $toIds, $locale)
     {
-        return $this->categoryRepository->findCategoryByKey($key);
+        $entities = $this->categoryRepository->findCategoriesBetween($fromIds, $toIds);
+
+        return $this->getApiObjects($entities, $locale);
     }
 
     /**
-     * Returns the categories with the given ids.
-     *
-     * @param array $ids
-     *
-     * @return CategoryEntity[]
+     * {@inheritdoc}
      */
-    public function findByIds(array $ids)
+    public function find($locale, $parentId = null)
     {
-        return $this->categoryRepository->findCategoryByIds($ids);
+        if ($parentId && !$this->categoryRepository->findCategoryById($parentId)) {
+            throw new CategoryNotFoundException($parentId);
+        }
+
+        $entities = $this->categoryRepository->findCategoriesByParentId($parentId);
+
+        return $this->getApiObjects($entities, $locale);
     }
 
     /**
-     * Creates a new category or overrides an existing one.
-     *
-     * @param array $data   The data of the category to save
-     * @param int   $userId The id of the user, who is doing this change
-     *
-     * @throws KeyNotUniqueException
-     *
-     * @return CategoryEntity
+     * {@inheritdoc}
      */
-    public function save($data, $userId)
+    public function save($data, $locale, $userId, $patch = false)
     {
-        try {
-            if (isset($data['id'])) {
-                return $this->modifyCategory($data, $this->getUser($userId));
+        $user = $this->userRepository->findUserById($userId);
+
+        if ($this->getProperty($data, 'id')) {
+            $categoryEntity = $this->findById($this->getProperty($data, 'id'), $locale)->getEntity();
+        } else {
+            $categoryEntity = new CategoryEntity();
+            $categoryEntity->setCreator($user);
+        }
+
+        $categoryEntity->setChanger($user);
+        $categoryWrapper = $this->getApiObject($categoryEntity, $locale);
+
+        if (!$patch || $this->getProperty($data, 'name')) {
+            $categoryWrapper->setName($this->getProperty($data, 'name', null));
+        }
+        if (!$patch || $this->getProperty($data, 'key')) {
```

Comment:
Would save `$this->getProperty($data, 'key')` in a separate variable, so that it is possible to reuse the value in the exception later.